atomix/cluster/src/main/java/io/atomix/raft/impl/ReconfigurationHelper.java

```diff
@@ -166,6 +170,170 @@ private void leaveInternal(final CompletableFuture<Void> future) {
+            // PASSIVE members yet. So we can safely assume they are all ACTIVE.
+            .map(m -> new DefaultRaftMember(m, Type.ACTIVE, Instant.now()))
+            .collect(Collectors.toSet());
+
+    if (!currentConfiguration.force()) {
+      // No need to overwrite if it is already in force configure and this is a retry
+      if (raftContext.getRaftRole().role() == Role.LEADER) {
+        // Optimization: If the current configuration is already the same as new forced, we
+        // can skip reconfiguring. It is most likely a retry of a previous force request,
+        // which was interpreted as failure because of a request timeout.
+        raftContext.transition(Role.FOLLOWER);
+      }
+
+      logger.info(
+          "Current configuration is '{}'. Forcing configuration with members '{}'",
+          currentConfiguration,
+          newMembers);
+      final var newConfiguration =
+          new Configuration(
+              currentConfiguration.index() + 1,
+              raftContext.getTerm(),
+              Instant.now().toEpochMilli(),
+              newMembers,
+              Set.of(),
+              true);
+
+      raftContext.getCluster().configure(newConfiguration);
+    } else if (!(currentConfiguration.allMembers().containsAll(newMembers)
+        && newMembers.containsAll(currentConfiguration.allMembers()))) {
+      // This is not expected. When force configuration is retried, we expect that they are
+      // retried with the same state. If this is not the case, it is likely that there are two
+      // force configuration requested at the same time.
+      // Reject the request. There is possibly no way out to recover from this.
+      future.completeExceptionally(
+          new IllegalStateException(
+              String.format(
+                  "Expected to force configure with members '%s', but the member is already in force configuration with a different set of members '%s'",
+                  newMembers, currentConfiguration.allMembers())));
+      return;
+    }
+
+    sendForceConfigureRequestToAllMembers(future);
+  }
+
+  private void sendForceConfigureRequestToAllMembers(final CompletableFuture<Void> future) {
+    final Configuration configuration = raftContext.getCluster().getConfiguration();
+    final var otherMembers =
+        configuration.newMembers().stream()
+            .map(RaftMember::memberId)
+            .filter(m -> !m.equals(raftContext.getCluster().getLocalMember().memberId()))
+            .collect(Collectors.toSet());
+
+    final var quorum =
+        new ForceConfigureQuorum(
+            success -> {
+              if (Boolean.TRUE.equals(success)) {
+                future.complete(null);
+              } else {
+                future.completeExceptionally(
+                    new ProtocolException(
+                        "Failed to force configure because not all members acknowledged the request."));
```

Comment:
:question: If this happens, I suspect we will want to know which members did not acknowledge the request, right? Is there any other way to find out right now?

Reply:
I suppose looking for the debug logs of success for the other members, and by elimination we find out which ones didn't acknowledge? :sweat_smile: I think it would be useful to have it here already in the message if we know.